File: src/matlab/com/springbok/pattern/PatternSRR_405V01Test.m
classdef PatternSRR_405V01Test < TestUtility
% Tests methods of PatternSRR_405V01 class.
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  properties (Constant = true)
    
    Phi0_input = 2;
    GainMax_input = 57;
    
    self = struct( ...
        'Phi0', 2);
    
    Phi_input = [0.2; 1; 5; 150];
    
    G_expected = [ ...
        56.880000000000003
        54.000000000000000
        27.000000000000000
        0
                 ];
    Gx_expected = [ ...
        18.830299622427006
        24.000000000000000
        9.956349637772750
        0
                  ];
    
  end % properties (Constant = true)
  
  methods
    
    function this = PatternSRR_405V01Test(logFId, testMode)
    % Constructs a PatternSRR_405V01Test
    %
    % Parameters:
    %   logFId - Log file identifier
    %   testMode - Test mode, if 'interactive' then beeps and pauses
      
    % Invoke superclass constructor
      if nargin == 0
        superArgs = {};
        
      else
        superArgs{1} = logFId;
        superArgs{2} = testMode;
        
      end % if
      this@TestUtility(superArgs{:});
      
    end % PatternSRR_405V01Test()
    
    function test_PatternSRR_405V01(this)
    % Tests PatternSRR_405V01 method.
      
      p = PatternSRR_405V01(this.Phi0_input);
      
      this.assert_true( ...
          'PatternSRR_405V01', ...
          'PatternSRR_405V01', ...
          this.HIGH_PRECISION_DESC, ...
          PatternUtility.compare_pattern(this.self, p));
      
    end % test_PatternSRR_405V01()
    
    function test_gain(this)
    % Tests gain method.
      
      p = PatternSRR_405V01(this.Phi0_input);
      
      [G_actual, Gx_actual] = p.gain(this.Phi_input, 'GainMax', this.GainMax_input);
      
      t(1) = max(abs(G_actual - this.G_expected) < this.HIGH_PRECISION);
      t(2) = max(abs(Gx_actual - this.Gx_expected) < this.HIGH_PRECISION);
      
      this.assert_true( ...
          'PatternSRR_405V01', ...
          'gain', ...
          this.HIGH_PRECISION_DESC, ...
          min(t));
      
      [G, Gx, fH] = p.gain([0:.001:180], 'GainMax', this.GainMax_input, 'PlotFlag', true);
      pause(1); close(fH);
      
      [G, Gx, fH] = p.gain([0:.001:180], 'PlotFlag', true);
      pause(1); close(fH);
      
    end % test_gain()
    
  end % methods
  
end % classdef

File: src/java/com/springbok/pattern/PatternSRR_405V01Test.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.pattern;

import com.springbok.utility.TestUtility;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertTrue;

/**
 * Tests methods of PatternSRR_405V01 class
 */
public class PatternSRR_405V01Test {
	private static final double Phi0_input = 2.0;
	private static final double GainMax_input = 57.0;

	private static final double[] Phi_input = new double[]{0.2, 1.0, 5.0, 150.0};
	private static final double[] G_expected = new double[]{56.880000000000003, 54.0,
			27.0, 0.0};
	private static final double[] Gx_expected = new double[]{18.830299622427006, 24.0,
			9.956349637772750, 0.0};

	/**
	 * Tests PatternSRR_405V01 method
	 */
	@Test
	public void test_PatternSRR_405V01() {
		PatternSRR_405V01 p = new PatternSRR_405V01(this.Phi0_input);

		assertTrue(TestUtility.isDoublesEquals(2.0, p.getPhi0()));
	}

	/**
	 * Tests gain method
	 */
	@Test
	public void test_gain() {
		PatternSRR_405V01 p = new PatternSRR_405V01(this.Phi0_input);

		Gain[] gains = new Gain[Phi_input.length];
		Map<String, Object> options = new HashMap<String, Object>();
		options.put("DoValidate", false);
		options.put("GainMax", GainMax_input);

		for (int i = 0; i < gains.length; i++) {
			Gain g = p.gain(Phi_input[i], options);
			gains[i] = g;
		}
		for (int i = 0; i < gains.length; i++) {
			assertTrue(TestUtility.isDoublesEquals(G_expected[i], gains[i].G));
			assertTrue(TestUtility.isDoublesEquals(Gx_expected[i], gains[i].Gx));
		}
	}
}
